pkg/syncer/upsync/upsync_process.go

```diff
@@ -0,0 +1,365 @@
+		if err != nil {
+			return err
+		}
+		_, err = downstreamNamespaceLister.Get(downstreamNamespace)
+		if k8serror.IsNotFound(err) {
+			// Downstream namespace not present; assume object is not present as well and remove upstream object
+			return c.removeUpstreamResource(ctx, gvr, clusterName, upstreamNamespace, upstreamName, true)
+		}
+		if err != nil {
+			return err
+		}
+
+		// Get namespaced downstream resource
+		downstreamObject, err = downstreamLister.ByNamespace(downstreamNamespace).Get(upstreamName)
+		if k8serror.IsNotFound(err) {
+			// Resource not found downstream prune it
+			return c.removeUpstreamResource(ctx, gvr, clusterName, upstreamNamespace, upstreamName, true)
+		}
+	} else {
+		// Get cluster-wide upstream resource
+		downstreamObject, err = downstreamLister.Get(upstreamName)
+		if k8serror.IsNotFound(err) {
+			// Resource not found downstream prune it
+			return c.removeUpstreamResource(ctx, gvr, clusterName, upstreamNamespace, upstreamName, true)
+		}
+	}
+
+	downstreamResource, ok := downstreamObject.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("type mismatch of resource object: received %T", downstreamResource)
+	}
+
+	if downstreamResource.GetDeletionTimestamp() != nil {
+		return c.removeUpstreamResource(ctx, gvr, clusterName, upstreamNamespace, upstreamName, false)
+	}
+
+	return err
+}
+
+func (c *controller) processDownstreamResource(ctx context.Context, gvr schema.GroupVersionResource, key string, includeStatus bool) error {
+	logger := klog.FromContext(ctx)
+	downstreamNamespace, downstreamName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Error(err, "Invalid key")
+		return nil
+	}
+
+	logger = logger.WithValues(DownstreamNamespace, downstreamNamespace, DownstreamName, downstreamName)
+
+	downstreamLister, err := c.getDownstreamLister(gvr)
+	if err != nil {
+		return err
+	}
+
+	var locatorHolder metav1.Object
+	if downstreamNamespace != "" {
+		downstreamNamespaceLister, err := c.getDownstreamLister(corev1.SchemeGroupVersion.WithResource("namespaces"))
+		if err != nil {
+			return err
+		}
+		nsObj, err := downstreamNamespaceLister.Get(downstreamNamespace)
+		if k8serror.IsNotFound(err) {
+			// Since the namespace is already deleted downstream we can delete the resource upstream as well
+			//
+			// TODO(davidfestal): the call below won't work afaict since  the key is a downstream one, not an upstream one.
+			// And in any case we don't have the locator to find out in where the upstream reosurce should be deleted.
+			//
+			// To manage this case we should catch downstream namespace delete events, in the resource handler,
+			// and get the upstream namespace (from the locator there, to then submit the deletion of the upstream value from there)
+			//
+			//			return c.pruneUpstreamResource(ctx, gvr, key)
+			//
+			logger.Error(err, "the downstream namespace doesn't exist anymore.")
```

Comment:
why is this an error?

Reply:
In full generality it is an error since we cannot cleanup the related synced upstream resource since we don't have the namespace locator to know in which upstream workspace/namespace to remove it.

Reply:
This might happen when a downstream resource is removed and the containing namespace are both deleted before the reconcilation happens.